apis/Google.Storage.V1/Google.Storage.V1/SignedUrlUtils.cs

```diff
@@ -0,0 +1,396 @@
+        /// </exception>
+        /// <returns>
+        /// The Signed URL which can be used to provide access to a bucket or object for a limited amount of time.
+        /// </returns>
+        public static string Create(
+            string bucket,
+            string objectName,
+            string credentialFilePath,
+            DateTimeOffset? expiration,
+            HttpMethod requestMethod = null,
+            Dictionary<string, IEnumerable<string>> requestHeaders = null,
+            Dictionary<string, IEnumerable<string>> contentHeaders = null)
+        {
+            StorageClientImpl.ValidateBucketName(bucket);
+            GaxPreconditions.CheckNotNull(credentialFilePath, nameof(credentialFilePath));
+
+            var credentials = ParseCredentials(credentialFilePath);
+            var timestamp = ((int?)((expiration - UnixEpoch)?.TotalSeconds))?.ToString();
+            var resourcePath = $"/{bucket}";
+            if (objectName != null)
+            {
+                resourcePath += $"/{objectName}";
+            }
+            var extensionHeaders = GatherExtensionHeaders(requestHeaders, contentHeaders);
+
+            var contentMD5 = GetFirstHeaderValue(contentHeaders, "Content-MD5");
+            var contentType = GetFirstHeaderValue(contentHeaders, "Content-Type");
+
+            var signatureLines = new List<string>
+            {
+                (requestMethod ?? HttpMethod.Get).ToString(),
+                contentMD5,
+                contentType,
+                timestamp
+            };
+            signatureLines.AddRange(extensionHeaders.Select(
+                header => $"{header.Key}:{string.Join(", ", header.Value)}"));
+            signatureLines.Add(resourcePath);
+
+            var signature = CreateSignature(string.Join("\n", signatureLines), credentials);
+
+            var queryParameters = new List<string> { $"GoogleAccessId={credentials.Id}" };
+            if (timestamp != null)
+            {
+                queryParameters.Add($"Expires={timestamp}");
+            }
+            queryParameters.Add($"Signature={signature}");
+            return $"{StorageHost}{resourcePath}?{string.Join("&", queryParameters)}";
+        }
+        
+        private static string Base64Encode(this byte[] value)
+        {
+            return Convert.ToBase64String(value);
+        }
+
+        private static string CreateSignature(string data, ServiceAccountCredential.Initializer initializer)
+        {
+            // TODO: This is taken from ServiceAccountCredential. Expose this somehow so we don't need to duplicate logic.
+            var hashAlg = SHA256.Create();
+            byte[] assertionHash = hashAlg.ComputeHash(Encoding.UTF8.GetBytes(data));
+#if NETSTANDARD1_3
+            var sigBytes = initializer.Key.SignHash(assertionHash, HashAlgorithmName.SHA256, RSASignaturePadding.Pkcs1);
+#else
+            const string Sha256Oig = "2.16.840.1.101.3.4.2.1";
```

Comment:
This was taken from ServiceAccountCredential, so its going away anyway when things get shared. I'd rather just leave as is for now to not clutter the code that will stay. As for the name, I'm not sure what it means exactly. I think its supposed to end with a 'd' actually instead of a 'g', because this is the [SHA256 OID](http://oid-info.com/get/2.16.840.1.101.3.4.2.1).